python/tank/bootstrap/resolver.py

```diff
@@ -131,36 +133,163 @@ def resolve_shotgun_configuration(
         :param sg_connection: Shotgun API instance
+        :param current_login: The login of the currently logged in user.
+
         :return: :class:`Configuration` instance
         """
         log.debug(
             "%s resolving configuration from Shotgun Pipeline Configuration %s" % (self, pipeline_config_name)
         )
 
-        log.warning(
-            "Shotgun pipeline configuration resolve has not been implemented yet. "
-            "Falling back on the base configuration resolver."
-        )
+        fields = [
+            "code",
+            "users",
+            "entry_point",
+            "sg_entry_point",
+            "windows_path",
+            "linux_path",
+            "mac_path",
+            "sg_descriptor",
+            "descriptor"
+        ]
+
+        pipeline_config = None
+
+        if pipeline_config_name is None:
+            log.debug("Will auto-detect which pipeline configuration to use.")
+
+            # get the pipeline configs for the current project which are
+            # either the primary or is associated with the currently logged in user.
+            pipeline_configs = sg_connection.find(
+                "PipelineConfiguration",
+                [{
+                    "filter_operator": "all",
+                    "filters": [
+                        ["project", "is", self._project_id],
+
+                        {
+                            "filter_operator": "any",
+                            "filters": [
+                                ["code", "is", constants.PRIMARY_PIPELINE_CONFIG_NAME],
+                                ["users.HumanUser.login", "contains", current_login]
+                            ]
+                        }
+                    ]
+                }],
+                fields,
+                order=[{"field_name": "updated_at", "direction": "asc"}]
+            )
+
+            log.debug(
+                "The following pipeline configurations were found: %s" % pprint.pformat(pipeline_configs)
+            )
+
+            # resolve primary and user config
+            primary_config = None
+            user_config = None
+            for pc in pipeline_configs:
+
+                # make sure configuration matches our entry point
+                if pc.get("entry_point") != self._entry_point and pc.get("sg_entry_point") != self._entry_point:
+                    continue
+
+                if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
+                    primary_config = pc
+                else:
+                    # user config
+                    if user_config:
+                        log.warning(
+                            "More than one user config detected. Will use the most "
+                            "recently updated one."
+                        )
+                    user_config = pc
+
+            # user pc takes precedence if available.
+            if user_config:
```

Comment:
minor: i like this explicitness, but I now others who would suggest reducing this to one line.